File: Android/app/src/main/java/uk/ewancroft/inkwell/ui/reader/IframeSecurityPolicy.kt
package uk.ewancroft.inkwell.ui.reader

import android.net.Uri

internal object IframeSecurityPolicy {
    fun isAllowedInitial(uri: Uri): Boolean =
        uri.scheme == "https" && !uri.host.isNullOrBlank()

    fun isAllowedNavigation(origin: Uri, candidate: Uri): Boolean =
        candidate.scheme == "https" &&
            candidate.host == origin.host &&
            effectivePort(candidate) == effectivePort(origin)

    private fun effectivePort(uri: Uri): Int = uri.port.takeIf { it != -1 } ?: 443
}

File: iOS/Inkwell/Features/Reader/IframeSecurityPolicy.swift
import Foundation

enum IframeSecurityPolicy {
    static func isAllowedInitial(_ url: URL) -> Bool {
        url.scheme?.lowercased() == "https" && url.host != nil
    }

    static func isAllowedNavigation(from origin: URL, to candidate: URL) -> Bool {
        guard isAllowedInitial(candidate),
              origin.scheme?.lowercased() == candidate.scheme?.lowercased(),
              origin.host?.lowercased() == candidate.host?.lowercased() else {
            return false
        }
        return effectivePort(origin) == effectivePort(candidate)
    }

    private static func effectivePort(_ url: URL) -> Int {
        url.port ?? 443
    }
}
